feat(cordyceps): added new push_back and pop_front methods to list

Tyrannican opened the pull request:
This PR adds some nice-to-have methods as detailed in [this issue](https://github.com/hawkw/mycelium/issues/186).

Changes:
* New `push_back` method
* New `pop_front` method
* Tests for each method

Small change but hope it's good :D

hawkw commented on cordyceps/src/list.rs:
```diff
@@ -158,6 +158,49 @@ impl<T: Linked<Links<T>> + ?Sized> List<T> {
+    /// Appends an item to the tail of the list
+    pub fn push_back(&mut self, item: T::Handle) {
+        let ptr = T::into_ptr(item);
+        // tracing::trace!(?self, ?ptr, "push_back");
```
my guess is that these are here because they're in the `push_back` and `pop_front` methods, but because they're commented out, i have a slight preference for not adding them in new code.

we should maybe eventually add a feature flag to control whether tracing is enabled in `cordyceps`... but we should probably do that in a separate branch.

Tyrannican replied:
Removed for this PR! :D

hawkw reviewed (commented):
thanks for the PR! i have a couple of very minor suggestions:

- the RustDoc for the `List` type references using the list as a FIFO using `push_front` and `pop_back`, it would be nice to add something saying that it can also be used as a stack or double-ended queue using these methods
- thanks for adding tests! it could be good to add these methods to the `proptest` `fuzz_linked_list` test as well, but if that's too much work, it would be fine to do that in a follow-up branch. 

thanks again for the PR!

Tyrannican commented:
* Added the tests to the `proptest` `fuzz_linked_list`
* Added documentation to detail the List data structure can be used as a stack or Doubly Linked List

hawkw reviewed (approved):
this looks great, i'm going to merge it once CI has run! thanks again!

hawkw reviewed (approved):
this looks great to me, i'm going to merge as soon as CI is done. thanks again!

hawkw commented:
oh, looks like `rustfmt` also needs to be run: https://github.com/hawkw/mycelium/runs/6747421224?check_suite_focus=true

mind fixing that? thanks!

Tyrannican commented:
Done!
